### DISCUSSION TITLE
unable to connect to perplexity api

### DISCUSSION THREAD

**Discussion by TheNasa** [open-webui/open-webui#3085] [General]
# Bug Report

## Description
unable to use perplexity api https://api.perplexity.ai
on perplexity docs
https://docs.perplexity.ai/docs/getting-started#:~:text=Our%20supported%20models%20are%20listed,easy%20integration%20with%20existing%20applications.
it said that its openai compatible
**Bug Summary:**
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.perplexity', port=443): Max retries exceeded with url: /models (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x7fe3ff085510>: Failed to resolve 'api.perplexity' ([Errno -2] Name or service not known)"))
INFO:     192.168.100.12:0 - "GET /openai/models/0 HTTP/1.1" 500 Internal Server Error

**Steps to Reproduce:**
[Outline the steps to reproduce the bug. Be as detailed as possible.]

**Expected Behavior:**
connected to perplexity api

**Actual Behavior:**
[error] OpenAI: Network Problem

## Environment

- **Open WebUI Version:** [e.g., 0.3.2]

- **Operating System:** [docker]

## Reproduction Details

**Confirmation:**

- [ ] I have read and followed all the instructions provided in the README.md.
- [ ] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ ] I have included the Docker container logs.

## Logs and Screenshots

**Browser Console Logs:**
[Include relevant browser console logs, if applicable]

**Docker Container Logs:**
[Include relevant Docker container logs, if applicable]

**Screenshots (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

## Installation Method

[Describe the method you used to install the project, e.g., manual installation, Docker, package manager, etc.]

## Additional Information

[Include any additional details that may help in understanding and reproducing the issue. This could include specific configurations, error messages, or anything else relevant to the bug.]

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by justinh-rahb** [comment]
It seems like the host being requested is `api.perplexity` and not `api.perplexity.ai`

**Comment by TheNasa** [comment]
> It seems like the host being requested is `api.perplexity` and not `api.perplexity.ai`


sorry that part of the log is when im trying with different url
this is the log for https://api.perplexity.ai/

INFO:config:Saving 'OPENAI_API_BASE_URLS' to config.json
INFO:     192.168.100.12:0 - "POST /openai/urls/update HTTP/1.1" 200 OK
INFO:apps.openai.main:get_all_models()
ERROR:apps.openai.main:Connection error: 0, message='Attempt to decode JSON with unexpected mimetype: text/plain; charset=utf-8', url=URL('https://api.perplexity.ai/models')
INFO:config:Saving 'OPENAI_API_KEYS' to config.json
INFO:     192.168.100.12:0 - "POST /openai/keys/update HTTP/1.1" 200 OK
INFO:apps.openai.main:get_all_models()
INFO:apps.openai.main:get_all_models()
ERROR:apps.openai.main:Connection error: 0, message='Attempt to decode JSON with unexpected mimetype: text/plain; charset=utf-8', url=URL('https://api.perplexity.ai/models')
ERROR:apps.openai.main:404 Client Error: Not Found for url: https://api.perplexity.ai/models
Traceback (most recent call last):
  File "/app/backend/apps/openai/main.py", line 322, in get_models
    r.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.perplexity.ai/models
ERROR:apps.openai.main:Connection error: 0, message='Attempt to decode JSON with unexpected mimetype: text/plain; charset=utf-8', url=URL('https://api.perplexity.ai/models')
INFO:     192.168.100.12:0 - "GET /openai/models/0 HTTP/1.1" 500 Internal Server Error
INFO:apps.openai.main:get_all_models()


on the perplexity docs, there is no mention of /models . but on open-webui /models is appended at the end of the url

**Comment by justinh-rahb** [comment]
you're missing `/v1`, the full endpoint URL should be:
```
https://api.perplexity.ai/v1
```

**Comment by TheNasa** [comment]
> you're missing `/v1`, the full endpoint URL should be:
> 
> ```
> https://api.perplexity.ai/v1
> ```

unfortunately i've tried https://api.perplexity.ai/v1 too and it didnt work


ERROR:apps.openai.main:404 Client Error: Not Found for url: https://api.perplexity.ai/v1/models
Traceback (most recent call last):
  File "/app/backend/apps/openai/main.py", line 322, in get_models
    r.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.perplexity.ai/v1/models
INFO:     10.5.5.3:0 - "GET /openai/models/0 HTTP/1.1" 500 Internal Server Error
INFO:apps.openai.main:get_all_models()

**Comment by justinh-rahb** [comment]
Turns out, Perplexity doesn't use a version in their API URL. It's possible that is what's interfering:
https://docs.perplexity.ai/discuss/6568121c0e55130062a5ee60

I haven't got a key to test it with, but attempting to hit that domain at all seems to return absolutely nothing as though it does not exist... have you gotten this to work with _anything else_?

**Reply by TheNasa** [reply]
it seems I'm able to connect using @Zambito1 method

**Comment by Zambito1** [comment]
FWIW I just got it to work by running a [Pipelines](https://github.com/open-webui/pipelines) server, and pasting https://raw.githubusercontent.com/open-webui/pipelines/main/examples/pipelines/providers/perplexity_manifold_pipeline.py into the Pipeline URL in http://localhost:3000/admin/settings/. That will ask for your API key, and then the Perplexity models will show up where your other models do.

**Reply by TheNasa** [reply]
> FWIW I just got it to work by running a [Pipelines](https://github.com/open-webui/pipelines) server, and pasting https://raw.githubusercontent.com/open-webui/pipelines/main/examples/pipelines/providers/perplexity_manifold_pipeline.py into the Pipeline URL in http://localhost:3000/admin/settings/. That will ask for your API key, and then the Perplexity models will show up where your other models do.

using your method works. Kudos!

**Reply by TheNasa** [reply]
I have another question. how does perplexity perform for you?
seems like mine does not remember any context. 
I'm not sure it is due to perplexity api or due to the pipeline itself.

**Reply by Zambito1** [reply]
You're right, it definitely does not maintain context. I would guess it is due to the pipeline. Probably because it uses `user_message` instead of `body`

https://github.com/open-webui/pipelines/blob/main/examples/pipelines/providers/perplexity_manifold_pipeline.py#L71

vs

https://github.com/open-webui/pipelines/blob/630e863fd3108e52947037cf8e7c431246cdd71a/examples/pipelines/providers/openai_pipeline.py#L55

**Reply by TheNasa** [reply]
hey, its me again. just want to let you know that i manage to get it running and maintain context. thank you so much for your help!

**Reply by Zambito1** [reply]
Did you make a new pipeline that passes the context?

**Reply by TheNasa** [reply]
yes, I make a new pipeline that appends each user message to a conversation history and includes this list in the payload to sent to the api

**Comment by k0gen** [comment]
I'll just add that verifying the connection (even if I specify the model) is failing for Perplexity. Despite this, I was able to talk to the `sonar` model just fine.

<img width="1437" alt="Screenshot 2025-02-18 at 16 57 59" src="https://github.com/user-attachments/assets/34eb1811-75e4-4538-8b41-cf6f0d672596" />
